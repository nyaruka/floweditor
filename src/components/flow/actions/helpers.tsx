import {
  Contact,
  Endpoints,
  Group,
  RecipientsAction,
  FlowIssue,
  FlowIssueType,
  BroadcastMsg,
  SendMsg
} from 'flowTypes';
import * as React from 'react';
import { Asset, AssetType } from 'store/flowContext';
import { FormEntry, NodeEditorSettings, ValidationFailure } from 'store/nodeEditor';
import { createUUID } from 'utils';
import { Trans } from 'react-i18next';
import shared from 'components/shared.module.scss';
import { showHelpArticle } from 'external';
import { IssueProps } from '../props';

export const renderIssues = (issueProps: IssueProps): JSX.Element => {
  const { issues, helpArticles } = issueProps;
  if (!issues || issues.length === 0) {
    return null;
  }

  return (
    <div style={{ padding: '10px 0px' }}>
      {issues.map((issue: FlowIssue, num: Number) => {
        const key = issue.node_uuid + issue.action_uuid + num;
        return (
          <div
            style={{ margin: '6px 0px', display: 'flex', fontSize: '14px', color: 'tomato' }}
            key={key}
          >
            <div
              style={{ marginRight: '8px', marginTop: '-2px', fontSize: '18px' }}
              className={`fe-warning`}
            />
            <div>{renderIssue(issue, helpArticles)}</div>
          </div>
        );
      })}
    </div>
  );
};

export const renderIssue = (
  issue: FlowIssue,
  helpArticles: { [key: string]: string } = {}
): JSX.Element => {
  // worst case, defer to the default description
  let message: JSX.Element = <>{issue.description}</>;

  if (issue.type === FlowIssueType.MISSING_DEPENDENCY) {
    message = (
      <Trans
        i18nKey="issues.missing_dependency"
        values={{
          name: issue.dependency.name || issue.dependency.key,
          type: issue.dependency.type
        }}
      >
        Cannot find a [[type]] for <span className="emphasize">[[name]]</span>
      </Trans>
    );
  }

  if (issue.type === FlowIssueType.INVALID_REGEX) {
    message = (
      <Trans i18nKey="issues.legacy_extra" values={{ regex: issue.regex }}>
        Invalid regular expression found: [[regex]]
      </Trans>
    );
  }

  if (issue.type === FlowIssueType.LEGACY_EXTRA) {
    message = (
      <Trans i18nKey="issues.legacy_extra">Expressions should not reference @legacy_extra</Trans>
    );
  }

  const article = helpArticles[issue.type];
  if (article) {
    return (
      <div
        className={shared.issue_help}
        onClick={() => {
          showHelpArticle(article);
        }}
      >
        {message}
      </div>
    );
  }

  return message;
};

export const getActionUUID = (nodeSettings: NodeEditorSettings, currentType: string): string => {
  if (nodeSettings.originalAction && nodeSettings.originalAction.type === currentType) {
    return nodeSettings.originalAction.uuid;
  }
  return createUUID();
};

export const getCompose = (action: SendMsg | BroadcastMsg = null): string => {
  if (!action) {
    return getEmptyComposeValue();
  }
  return action.compose;
};

export const getEmptyComposeValue = (): string => {
  return JSON.stringify({ text: '', attachments: [] });
};

export const getRecipients = (action: RecipientsAction): Asset[] => {
  let selected: any[] = (action.groups || []).map((group: Group) => {
    return {
      id: group.uuid,
      name: group.name,
      type: AssetType.Group
    };
  });

  selected = selected.concat(
    (action.contacts || []).map((contact: Contact) => {
      return { id: contact.uuid, name: contact.name, type: AssetType.Contact, missing: false };
    })
  );

  selected = selected.concat(
    (action.legacy_vars || []).map((expression: string) => {
      return { name: expression, value: expression, expression: true };
    })
  );

  return selected;
};

export const renderAssetList = (
  assets: Asset[],
  max: number = 10,
  endpoints: Endpoints
): JSX.Element[] => {
  // show our missing ones first
  return assets.reduce((elements, asset, idx) => {
    if (idx <= max - 2 || assets.length === max) {
      elements.push(renderAsset(asset, endpoints));
    } else if (idx === max - 1) {
      elements.push(<div key="ellipses">+{assets.length - max + 1} more</div>);
    }
    return elements;
  }, []);
};

export const renderAsset = (asset: Asset, endpoints: Endpoints) => {
  let assetBody = null;

  switch (asset.type) {
    case AssetType.Classifier:
      assetBody = (
        <Trans i18nKey="assets.classifier" values={{ name: asset.name }}>
          Call [[name]] classifier
        </Trans>
      );
      break;
    case AssetType.Group:
      assetBody = (
        <>
          <span className={`${shared.node_group} fe-group`} />
          {asset.name}
        </>
      );
      break;
    case AssetType.Label:
      assetBody = (
        <>
          <span className={`${shared.node_label} fe-label`} />
          {asset.name}
        </>
      );
      break;
    case AssetType.Flow:
      assetBody = (
        <>
          <span className={`${shared.node_label} fe-split`} />
          <a
            onMouseDown={(e: any) => {
              e.preventDefault();
              e.stopPropagation();
            }}
            onMouseUp={(e: any) => {
              e.preventDefault();
              e.stopPropagation();
            }}
            href={`${endpoints.editor}/${asset.id}`}
            rel="noopener noreferrer"
            target="_blank"
          >
            {asset.name}
          </a>
        </>
      );
      break;
    case AssetType.Ticketer:
      assetBody = (
        <Trans i18nKey="assets.ticketer" values={{ name: asset.name }}>
          Open a new Ticket on [[name]]
        </Trans>
      );
      break;
  }

  if (!assetBody) {
    assetBody = asset.name;
  }

  return (
    <div className={`${shared.node_asset}`} key={asset.id || (asset as any).value}>
      {assetBody}
    </div>
  );
};

export const getAllErrors = (entry: FormEntry): ValidationFailure[] => {
  return entry ? entry.validationFailures || [] : [];
};

export const hasErrors = (entry: FormEntry): boolean => {
  return getAllErrors(entry).length > 0;
};

export const getAllErrorMessages = (entry: FormEntry): string[] => {
  const errors = getAllErrors(entry).map((failure: ValidationFailure) => failure.message);
  return errors;
};

export const getExpressions = (assets: any[]): any[] => {
  return assets
    .filter((asset: any) => asset.expression)
    .map((asset: any) => {
      return asset.value;
    });
};

export const getRecipientsByAsset = (assets: Asset[], type: AssetType): any[] => {
  return assets
    .filter((asset: Asset) => asset.type === type)
    .map((asset: Asset) => {
      return { uuid: asset.id, name: asset.name };
    });
};

export const getComposeByAsset = (value: string, asset: string): any | any[] => {
  return JSON.parse(value)[asset];
};
