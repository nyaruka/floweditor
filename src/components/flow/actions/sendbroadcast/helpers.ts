import {
  getActionUUID,
  getComposeActionToState,
  getComposeStateToAction,
  getRecipients,
  getRecipientsStateToAction
} from 'components/flow/actions/helpers';
import { SendBroadcastFormState } from 'components/flow/actions/sendbroadcast/SendBroadcastForm';
import { Types } from 'config/interfaces';
import { BroadcastMsg } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): SendBroadcastFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.send_broadcast) {
    let action = settings.originalAction as BroadcastMsg;

    // check if our form should use a localized action
    if (settings.localizations && settings.localizations.length > 0) {
      const localized = settings.localizations[0];
      if (localized.isLocalized()) {
        action = settings.localizations[0].getObject() as BroadcastMsg;
      } else {
        return {
          compose: { value: getComposeActionToState() },
          recipients: { value: [] },
          valid: true
        };
      }
    }

    return {
      compose: { value: getComposeActionToState(action) },
      recipients: { value: getRecipients(action) },
      valid: true
    };
  }

  return {
    compose: { value: getComposeActionToState() },
    recipients: { value: [] },
    valid: false
  };
};

export const stateToAction = (
  settings: NodeEditorSettings,
  formState: SendBroadcastFormState
): BroadcastMsg => {
  const [compose, text, attachments] = getComposeStateToAction(formState);
  const [legacy_vars, contacts, groups] = getRecipientsStateToAction(formState);

  return {
    legacy_vars: legacy_vars,
    contacts: contacts,
    groups: groups,
    compose: compose,
    text: text,
    attachments: attachments,
    type: Types.send_broadcast,
    uuid: getActionUUID(settings, Types.send_broadcast)
  };
};
