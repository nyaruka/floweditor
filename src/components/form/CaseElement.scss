@import '../../variables.scss';

$spacing: 10px;

.kase {
    width: 100%;
    height: 25px;
    padding-top: 15px;
    .form-group {
        margin: 0;
    }
    .choice {
        width: 185px;
        display: inline-block;
    }
    .operand {
        display: inline-block;
        position: relative;
        top: -10px;
        margin-left: $spacing;
        width: 166px;
        input {
            width: 166px;
        }
    }
    .categorize-as {
        display: inline-block;
        font-size: 13px;
        line-height: 13px;
        position: relative;
        top: -10px;
        margin: 0px $spacing;
    }
    .category {
        display: inline-block;
        position: relative;
        top: -10px;
        input {
            width: 105px;
        }
    }
    &:hover {
        .remove-icon {
            opacity: 1;
        }

        .dnd-icon {
            opacity: 1;
        }
    }
    input {
        padding: 6px 7px;
        height: 28px !important;
        font-size: 13px !important;
        border: 1px solid #ccc;
    }
}

.dnd-icon {
    color: #ccc;
    font-weight: bold;
    padding-top: 8px;
    padding-right: 5px;
    float: left;
    opacity: 0;
    font-size: 18px;
    line-height: 12px;
}

.remove-icon {
    color: #ccc;
    padding-top: 5px;
    float: right;
    opacity: 0;
    font-size: 16px;
    line-height: 12px;
}
