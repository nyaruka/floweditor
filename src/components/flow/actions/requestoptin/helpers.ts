import { getActionUUID } from 'components/flow/actions/helpers';
import { Types } from 'config/interfaces';
import { RequestOptIn } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';
import { RequestOptInFormState } from './RequestOptInForm';

export const initializeForm = (settings: NodeEditorSettings): RequestOptInFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.request_optin) {
    const action = settings.originalAction as RequestOptIn;
    return {
      optin: { value: action.optin },
      valid: true
    };
  }

  return {
    optin: { value: null },
    valid: false
  };
};

export const stateToAction = (
  settings: NodeEditorSettings,
  formState: RequestOptInFormState
): RequestOptIn => {
  const result = {
    type: Types.request_optin,
    optin: formState.optin.value[0],
    uuid: getActionUUID(settings, Types.add_input_labels)
  };
  return result;
};
