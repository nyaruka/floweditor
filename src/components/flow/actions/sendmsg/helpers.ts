/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { getActionUUID } from 'components/flow/actions/helpers';
import { SendMsgFormState } from 'components/flow/actions/sendmsg/SendMsgForm';
import { Types } from 'config/interfaces';
import { FlowEditorConfig, Label, MsgTemplating, SendMsg } from 'flowTypes';
import { FormEntry, NodeEditorSettings, StringEntry } from 'store/nodeEditor';
import { SelectOption } from 'components/form/select/SelectElement';
import { createUUID } from 'utils';
import { Attachment } from './attachments';

export const TOPIC_OPTIONS: SelectOption[] = [
  { value: 'event', name: 'Event' },
  { value: 'account', name: 'Account' },
  { value: 'purchase', name: 'Purchase' },
  { value: 'agent', name: 'Agent' }
];

export const initializeForm = (
  settings: NodeEditorSettings,
  config: FlowEditorConfig
): SendMsgFormState => {
  let template: FormEntry = { value: null };
  let templateVariables: StringEntry[] = [];
  let skipValidation = false;

  if (settings.originalAction && settings.originalAction.type === Types.send_msg) {
    const action = settings.originalAction as SendMsg;
    const attachments: Attachment[] = [];
    let expressionValue = null;
    (action.attachments || []).forEach((attachmentString: string) => {
      const splitPoint = attachmentString.indexOf(':');

      const type = attachmentString.substring(0, splitPoint);
      const attachment = {
        type,
        url: attachmentString.substring(splitPoint + 1),
        uploaded: type.indexOf('/') > -1,
        valid: false
      };

      attachments.push(attachment);
    });

    if (action.templating) {
      const msgTemplate = action.templating.template;
      if (action.templating.expression) {
        expressionValue = { value: action.templating.expression };
      }
      template = {
        value: {
          uuid: msgTemplate.uuid,
          name: msgTemplate.name
        }
      };
      templateVariables = action.templating.variables.map((value: string) => {
        return {
          value
        };
      });
    }

    const labels = action.labels
      ? action.labels.map((label: Label) => {
          if (label.name_match) {
            return { name: label.name_match, expression: true };
          }
          return label;
        })
      : [];

    if (config.skipValidation) {
      skipValidation = config.skipValidation;
    }

    return {
      expression: expressionValue,
      topic: { value: TOPIC_OPTIONS.find(option => option.value === action.topic) },
      template,
      templateVariables,
      attachments,
      labels: {
        value: labels
      },
      uploadInProgress: false,
      uploadError: '',
      message: { value: action.text },
      quickReplies: { value: action.quick_replies || [] },
      quickReplyEntry: { value: '' },
      sendAll: action.all_urns,
      valid: true,
      skipValidation
    };
  }

  return {
    topic: { value: null },
    template,
    templateVariables: [],
    attachments: [],
    uploadInProgress: false,
    uploadError: '',
    message: { value: '' },
    quickReplies: { value: [] },
    quickReplyEntry: { value: '' },
    sendAll: false,
    valid: false,
    labels: { value: [] }
  };
};

export const stateToAction = (settings: NodeEditorSettings, state: SendMsgFormState): SendMsg => {
  const attachments = state.attachments
    .filter((attachment: Attachment) => attachment.url.trim().length > 0)
    .map((attachment: Attachment) => `${attachment.type}:${attachment.url}`);

  let templating: MsgTemplating = null;

  if (state.template && state.template.value) {
    let templatingUUID = createUUID();
    if (settings.originalAction && settings.originalAction.type === Types.send_msg) {
      const action = settings.originalAction as SendMsg;
      if (
        action.templating &&
        action.templating.template &&
        action.templating.template.uuid === state.template.value.id
      ) {
        templatingUUID = action.templating.uuid;
      }
    }

    templating = {
      uuid: templatingUUID,
      template: {
        uuid: state.template.value.uuid,
        name: state.template.value.name
      },
      variables: state.templateVariables.map((variable: StringEntry) => variable.value)
    };

    if (state.expression && state.expression.value) {
      templating.expression = state.expression.value;
    }
  }
  const result: SendMsg = {
    attachments,
    text: state.message.value,
    type: Types.send_msg,
    all_urns: state.sendAll,
    labels: state.labels.value.map((label: any) => {
      if (label.expression) {
        return { name_match: label.name };
      }
      return label;
    }),
    quick_replies: state.quickReplies.value,
    uuid: getActionUUID(settings, Types.send_msg)
  };

  if (templating) {
    result.templating = templating;
  }

  if (state.topic.value) {
    result.topic = state.topic.value.value;
  }

  return result;
};
