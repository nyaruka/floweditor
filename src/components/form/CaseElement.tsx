import * as React from 'react';
import Select from 'react-select';
import { v4 as generateUUID } from 'uuid';
import { Operator } from '../../providers/ConfigProvider/operatorConfigs';
import ComponentMap from '../../services/ComponentMap';
import FormElement from './FormElement';
import TextInputElement, { HTMLTextElement } from './TextInputElement';
import {
    operatorConfigListPT,
    getOperatorConfigPT
} from '../../providers/ConfigProvider/propTypes';
import { ConfigProviderContext } from '../../providers/ConfigProvider/configContext';
import { Case } from '../../flowTypes';
import { ChangedCaseInput } from '../routers/SwitchRouter';

import * as forms from './FormElement.scss';
import * as styles from './CaseElement.scss';

export interface CaseElementProps {
    name?: string; // satisfy form widget props
    onRemove?(c: CaseElement): void;
    ComponentMap?: ComponentMap;
    kase: Case;
    exitName: string;
    empty?: boolean;
    onChanged?(c: any, type?: ChangedCaseInput): void;
    focusArgsInput?: boolean;
    focusExitInput?: boolean;
    solo?: boolean;
}

interface CaseElementState {
    errors: string[];
    operator: string;
    arguments: string[];
    exitName: string;
}

export const prefix = (operatorType: string): string => {
    let pre = '';
    if (operatorType.indexOf('_lt') > -1) {
        if (operatorType.indexOf('date') > -1) {
            pre = 'Before ';
        } else {
            if (operatorType.indexOf('lte') > -1) {
                pre = '<= ';
            } else {
                pre = '< ';
            }
        }
    } else if (operatorType.indexOf('_gt') > -1) {
        if (operatorType.indexOf('date') > -1) {
            pre = 'After ';
        } else {
            if (operatorType.indexOf('gte') > -1) {
                pre = '>= ';
            } else {
                pre = '>';
            }
        }
    }
    return pre;
};

export const composeExitName = (operator: string, newArgList: string[]): string => {
    const pre: string = prefix(operator);
    if (newArgList.length > 0) {
        const [firstArg] = newArgList;
        const words = firstArg.match(/\w+/g);
        if (words && words.length > 0) {
            const [firstWord] = words;
            return pre + firstWord.charAt(0).toUpperCase() + firstWord.slice(1);
        }
        return pre + firstArg.charAt(0).toUpperCase() + firstArg.slice(1);
    }
    return pre;
};

export const getExitName = (
    exitName: string,
    operator: string,
    operatorConfig: Operator,
    kase: Case,
    newArgList: string[] = []
): string => {
    // Don't reassign args
    let newExitName = exitName;
    // Some operators don't expect args
    if (newArgList.length) {
        if (!newExitName || newExitName === composeExitName(operator, kase.arguments)) {
            newExitName = composeExitName(operator, newArgList);
        }
    } else {
        // If the operator has a default category name, use that
        if (operatorConfig.categoryName) {
            ({ categoryName: newExitName } = operatorConfig);
        }
    }
    return newExitName;
};

export default class CaseElement extends React.Component<CaseElementProps, CaseElementState> {
    private category: TextInputElement;

    public static contextTypes = {
        operatorConfigList: operatorConfigListPT,
        getOperatorConfig: getOperatorConfigPT
    };

    public state: CaseElementState = {
        errors: [],
        operator: this.props.kase.type,
        arguments: this.props.kase.arguments,
        exitName: this.props.exitName ? this.props.exitName : ''
    };

    private operatorConfig: Operator = this.context.getOperatorConfig(this.props.kase.type);

    private onChangeOperator = (val: Operator): void => {
        this.operatorConfig = val;

        this.setState(
            {
                operator: val.type,
                exitName: getExitName(
                    this.state.exitName,
                    this.state.operator,
                    this.operatorConfig,
                    this.props.kase
                )
            },
            () => this.props.onChanged(this)
        );
    };

    private onChangeArguments = (val: React.ChangeEvent<HTMLTextElement>): void => {
        const args = [val.target.value];
        const exitName = getExitName(
            this.state.exitName,
            this.state.operator,
            this.operatorConfig,
            this.props.kase,
            args
        );

        // prettier-ignore
        this.setState(
            {
                arguments: args,
                exitName
            },
            () => {
                this.props.onChanged(this, ChangedCaseInput.ARGS);
                this.category.setState(
                    {
                        value: exitName
                    },
                    () => {
                        /** If the case doesn't have both an argument & an exit name, remove it */
                        if (!this.state.arguments[0] && !this.state.exitName) {
                            return this.remove();
                        }
                    }
                );
            }
        );
    };

    private onChangeExitName = (val: React.ChangeEvent<HTMLTextElement>): void =>
        this.setState(
            {
                exitName: val.target.value
            },
            () => this.props.onChanged(this, ChangedCaseInput.EXIT)
        );

    private remove = (ele?: any): void => this.props.onRemove(this);

    private hasArguments = (): boolean =>
        this.state.arguments &&
        this.state.arguments.length > 0 &&
        this.state.arguments[0].trim().length > 0;

    public validate(): boolean {
        const errors: string[] = [];

        if (this.operatorConfig.operands === 0) {
            if (this.state.exitName.trim().length === 0) {
                const { verboseName } = this.operatorConfig;
                errors.push(`A category name is required when using "${verboseName}"`);
            }
        } else {
            /**
             * Check our argument list
             * If we have arguments, we need an exit name
             */
            if (this.hasArguments()) {
                if (!this.category || !this.category.state.value) {
                    errors.push('A category name is required');
                }
            }

            /** If we have an exit name we need arguments */
            if (this.state.exitName) {
                if (!this.hasArguments()) {
                    const operator = this.context.getOperatorConfig(this.state.operator);
                    const { verboseName } = operator;
                    errors.push(`When using "${verboseName}", an argument is required.`);
                }
            }

            /** Validate numeric and date operators */
            if (this.hasArguments() && this.state.arguments[0].trim().indexOf('@') !== 0) {
                if (this.state.operator.indexOf('number') > -1) {
                    if (this.state.arguments[0]) {
                        if (isNaN(parseInt(this.state.arguments[0], 10))) {
                            errors.push('Enter a number when using numeric rules.');
                        }
                    }
                }

                if (this.state.operator.indexOf('date') > -1) {
                    if (this.state.arguments[0]) {
                        if (isNaN(Date.parse(this.state.arguments[0]))) {
                            errors.push('Enter a date when using date rules (e.g. 1/1/2017).');
                        }
                    }
                }
            }
        }

        if (this.state.errors.length === 0 && errors.length === 0) {
            return true;
        }

        this.setState({ errors });

        return errors.length === 0;
    }

    private getArgs(): JSX.Element {
        let args: JSX.Element = null;

        if (this.operatorConfig && this.operatorConfig.operands > 0) {
            const value = this.state.arguments ? this.state.arguments[0] : '';

            args = (
                <TextInputElement
                    data-spec="args-input"
                    name="arguments"
                    onChange={this.onChangeArguments}
                    value={value}
                    focus={this.props.focusArgsInput}
                    autocomplete={true}
                    ComponentMap={this.props.ComponentMap}
                />
            );
        }

        return args;
    }

    private getDndIco(): JSX.Element {
        let dndIco: JSX.Element = null;

        if (!this.props.empty && !this.props.solo) {
            dndIco = (
                <div className={styles.dndIcon}>
                    <span>&#8597;</span>
                </div>
            );
        } else {
            dndIco = <div style={{ display: 'inline-block', width: 15 }} />;
        }

        return dndIco;
    }

    private getRemoveIco(): JSX.Element {
        let removeButton: JSX.Element = null;

        if (!this.props.empty) {
            removeButton = (
                <div className={styles.removeButton} onClick={this.remove}>
                    <span className="icon-remove" />
                </div>
            );
        }

        return removeButton;
    }

    public render(): JSX.Element {
        const args: JSX.Element = this.getArgs();
        const dndIco: JSX.Element = this.getDndIco();
        const removeIco: JSX.Element = this.getRemoveIco();

        return (
            <FormElement
                data-spec="case-form"
                name={this.props.name}
                errors={this.state.errors}
                className={styles.kase}
                case={true}>
                <div className="select-medium">
                    {dndIco}
                    <div className={styles.choice}>
                        <Select
                            data-spec="operator-list"
                            name="operator"
                            clearable={false}
                            options={this.context.operatorConfigList}
                            value={this.state.operator}
                            valueKey="type"
                            labelKey="verboseName"
                            optionClassName="operator"
                            searchable={false}
                            onChange={this.onChangeOperator}
                        />
                    </div>
                    <div className={styles.operand}>{args}</div>
                    <div className={styles.categorizeAs}>categorize as</div>
                    <div className={styles.category}>
                        <TextInputElement
                            ref={ele => (this.category = ele)}
                            data-spec="exit-input"
                            name="exitName"
                            onChange={this.onChangeExitName}
                            value={this.state.exitName}
                            focus={this.props.focusExitInput}
                            ComponentMap={this.props.ComponentMap}
                        />
                    </div>
                    {removeIco}
                </div>
            </FormElement>
        );
    }
}
