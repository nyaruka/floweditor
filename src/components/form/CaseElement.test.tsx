import * as React from 'react';
import { shallow, mount } from 'enzyme';
import CompMap from '../../services/ComponentMap';
import CaseElement, { prefix, composeExitName, getExitName, hasArgs } from './CaseElement';
import { getSpecWrapper, titleCase } from '../../helpers/utils';
import { object } from 'prop-types';
import {
    operatorConfigList,
    getOperatorConfig
} from '../../providers/ConfigProvider/operatorConfigs';

const {
    results: [{ definition }]
} = require('../../../test_flows/a4f64f1b-85bc-477e-b706-de313a022979.json');

const { nodes: [, node] } = definition;
const { router: { cases }, exits } = node;

const context = {
    operatorConfigList,
    getOperatorConfig
};

const ComponentMap = new CompMap(definition);

describe('CaseElement >', () => {
    describe('helpers >', () => {
        describe('prefix >', () =>
            operatorConfigList.forEach(({ verboseName, type }) =>
                it(`should prefix "${verboseName}" operator appropriately`, () =>
                    expect(prefix(type)).toMatchSnapshot())
            ));

        describe('composeExitName >', () => {
            const operatorType = 'has_any_word';
            const strArgOperators = operatorConfigList.slice(0, 6);
            let args = [['tomato, t'], ['papayawhip'], []];

            it('should handle empty arg lists appropriately', () =>
                expect(composeExitName(operatorType, args[2])).toBe(''));

            it('should return the first arg in list, capitalized', () => {
                args = args.slice(0, 2);
                strArgOperators.forEach(({ type }) =>
                    args.slice(0, 2).forEach(argList => {
                        const [firstArg] = argList[0].split(',');
                        expect(composeExitName(type, argList)).toBe(titleCase(firstArg));
                    })
                );
            });
        });

        describe('getExitName >', () => {
            const type = 'has_value';
            const kase = {
                uuid: 'fa0a9b24-5f19-4b8e-b287-27af5811de1d',
                type,
                exit_uuid: '55855afc-f612-4ef9-9288-dcb1dd136052'
            };
            const operatorConfig = getOperatorConfig(type);
            const { categoryName: expectedExitName } = operatorConfig;

            it("should return a given operator's default category name if it has one", () =>
                // prettier-ignore
                expect(
                    getExitName(
                        'somePerhapsPreExistingExitName',
                        operatorConfig,
                        kase
                    )
                ).toBe(expectedExitName));
        });

        describe('hasArgs >', () =>
            it(`should reflect the presence of an arguments array in the passed object's "arguments" property`, () => {
                expect(hasArgs(['tomato, t'])).toBeTruthy();
                expect(hasArgs([])).toBeFalsy();
                expect(hasArgs()).toBeFalsy();
            }));
    });

    describe('Component >', () => {
        const caseUUID = '29b18c7e-c232-414c-9fc0-2e0b6260d9ca';
        const onChange = jest.fn();
        const onRemove = jest.fn();
        const props = {
            name: `case_${caseUUID}`,
            kase: {
                uuid: caseUUID,
                type: 'has_any_word',
                exit_uuid: '38c1m4g4-b424-585d-8cgi-384d6260ymca'
            },
            exitName: 'Red, ',
            empty: true,
            onRemove,
            onChange,
            focusArgsInput: false,
            focusExitInput: false,
            ComponentMap
        };

        describe('render >', () => {
            it('should render empty case', () => {
                const EmptyCase = shallow(<CaseElement {...{ ...props, exitName: null }} />, {
                    context
                });

                const {
                    onChangeOperator,
                    onChangeArguments,
                    onChangeExitName
                } = EmptyCase.instance() as any;

                expect(getSpecWrapper(EmptyCase, 'case-form').props()).toEqual(
                    expect.objectContaining({
                        name: props.name,
                        errors: [],
                        case: true
                    })
                );

                expect(getSpecWrapper(EmptyCase, 'operator-list').props()).toEqual(
                    expect.objectContaining({
                        options: operatorConfigList,
                        clearable: false,
                        name: 'operator',
                        optionClassName: 'operator',
                        labelKey: 'verboseName',
                        valueKey: 'type',
                        value: props.kase.type,
                        'data-spec': 'operator-list',
                        onChange: onChangeOperator
                    })
                );

                expect(getSpecWrapper(EmptyCase, 'args-input').props()).toEqual(
                    expect.objectContaining({
                        'data-spec': 'args-input',
                        name: 'arguments',
                        onChange: onChangeArguments,
                        value: '',
                        focus: props.focusArgsInput,
                        autocomplete: true,
                        ComponentMap
                    })
                );

                expect(getSpecWrapper(EmptyCase, 'exit-input').props()).toEqual({
                    'data-spec': 'exit-input',
                    name: 'exitName',
                    onChange: onChangeExitName,
                    value: '',
                    focus: props.focusExitInput,
                    ComponentMap
                });
            });

            cases.forEach((kase, idx) => {
                const caseProps = {
                    name: `case_${kase.uuid}`,
                    kase,
                    exitName: exits[idx].name,
                    onRemove: jest.fn(),
                    onChange: jest.fn(),
                    focusArgsInput: false,
                    focusExitInput: false,
                    ComponentMap
                };
                const CaseWrapper = shallow(<CaseElement {...caseProps} />, {
                    context
                });

                it('should render FormElements with expected props', () =>
                    expect(getSpecWrapper(CaseWrapper, 'case-form').props()).toEqual(
                        expect.objectContaining({
                            name: caseProps.name,
                            errors: []
                        })
                    ));

                it('should render operator lists w/ expected props', () =>
                    expect(getSpecWrapper(CaseWrapper, 'operator-list').props()).toEqual(
                        expect.objectContaining({
                            value: kase.type
                        })
                    ));

                it('should render arguments inputs w/ expected props', () =>
                    expect(getSpecWrapper(CaseWrapper, 'args-input').props()).toEqual(
                        expect.objectContaining({
                            value:
                                kase.arguments.constructor === Array
                                    ? kase.arguments[0]
                                    : kase.arguments
                        })
                    ));

                it('should render exit inputs w/ expected props', () =>
                    expect(getSpecWrapper(CaseWrapper, 'exit-input').props()).toEqual(
                        expect.objectContaining({
                            value: caseProps.exitName || '',
                            focus: caseProps.focusExitInput
                        })
                    ));
            });
        });

        describe('instance methods >', () => {
            describe('onChangeOperator >', () => {
                const Case = mount(<CaseElement {...props} />, { context });
                const { onChangeOperator } = Case.instance() as any;
                const [anyWordOperator, allWordsOperator] = operatorConfigList;

                it("should update state operator config to the value it's passed if value is different than operatorConfig in state", () => {
                    onChangeOperator(allWordsOperator);
                    expect(Case.state('operatorConfig')).toEqual(allWordsOperator);
                });

                it("should update exitName if value it's passed is different than operatorConfig in state", () => {
                    onChangeOperator(anyWordOperator);
                    expect(Case.state('exitName')).toBe(
                        getExitName(
                            Case.state('exitName'),
                            anyWordOperator,
                            props.kase,
                            Case.state('arguments')
                        )
                    );
                });

                it('should call "onChange" prop after setting own state', () => {
                    onChangeOperator(allWordsOperator);
                    expect(onChange).toHaveBeenCalled();
                });
            });
        });
    });
});
