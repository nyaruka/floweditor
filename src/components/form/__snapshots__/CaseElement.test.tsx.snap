// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CaseElement utility methods prefix should prefix 'has a date after' operator accordingly 1`] = `"After "`;

exports[`Component: CaseElement utility methods prefix should prefix 'has a date before' operator accordingly 1`] = `"Before "`;

exports[`Component: CaseElement utility methods prefix should prefix 'has a date equal to' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has a date' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has a number above' operator accordingly 1`] = `">"`;

exports[`Component: CaseElement utility methods prefix should prefix 'has a number at or above' operator accordingly 1`] = `">= "`;

exports[`Component: CaseElement utility methods prefix should prefix 'has a number at or below' operator accordingly 1`] = `"<= "`;

exports[`Component: CaseElement utility methods prefix should prefix 'has a number below' operator accordingly 1`] = `"< "`;

exports[`Component: CaseElement utility methods prefix should prefix 'has a number between' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has a number equal to' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has a number' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has a phone number' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has a run status of' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has all of the words' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has an email' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has an error' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has any of the words' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has only the phrase' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has some text' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'has the phrase' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'is in the group' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'is not empty' operator accordingly 1`] = `""`;

exports[`Component: CaseElement utility methods prefix should prefix 'starts with' operator accordingly 1`] = `""`;
